x-pack/elastic-agent/pkg/core/server/server.go

```diff
@@ -548,8 +549,10 @@ func (as *ApplicationState) Stop(timeout time.Duration) error {
 		s := as.status
 		doneChan := as.checkinDone
 		as.checkinLock.RUnlock()
-		if s == proto.StateObserved_STOPPING && doneChan == nil {
-			// sent stopping and now is disconnected (so its stopped)
+		if (wasConn && doneChan == nil) || (!wasConn && s == proto.StateObserved_STOPPING && doneChan == nil) {
```

Comment:
i'm not sure i follow this. if it was connected and doneChan is nil means it got disconnected this seems ok.
second part means if status of application is stopping and doneChan is nil (got disconnected) then we;re destroying but only in case doneChan was nil before so nothing changed.

Reply:
Yes the second case is only in the case that `Stop()` was called but the client was disconnected from the GRPC at that time (which is very rare, but possible).

So if the client was disconnected from the GPRC at the time `Stop()` was called, it needs to know that it did receive the stopping state. So it waits for the client to send that it is actually stopping and then it has disconnected. This requires that the client actually reconnect to get the stopping message or timeout occurs, which ever comes first.

In the normal case the `wasConn && doneChan == nil` will almost always be used in this loop.

Hopefully that it explains it better.

Reply:
makes sense thanks